python/mxnet/optimizer.py

```diff
@@ -670,9 +670,10 @@ def __init__(self, optimizer):
 
     def __call__(self, index, grad, weight):
         """Update weight given gradient and index"""
-        if index not in self.states:
-            self.states[index] = self.optimizer.create_state(index, weight)
-        self.optimizer.update(index, weight, grad, self.states[index])
+        key = (index, str(weight.context))
```

Comment:
You mean from a checkpoint? Could you point me to where I can look to fix it?

Since I'm using reduce-scatter to allgather, I have shards of data distributed across all participating gpus. This requires state like momentum to also be sharded across gpus. Otherwise there is either slowdown, when one GPU needs to access the other GPU memory to apply it, or crash if there is no P2P connection between them.